fastapi/concurrency.py

```diff
@@ -24,7 +30,10 @@ async def contextmanager_in_threadpool(
     # works (1 is arbitrary)
     exit_limiter = CapacityLimiter(1)
     try:
-        yield await run_in_threadpool(cm.__enter__)
+        # Semaphore ensures there's always one thread left to call __exit__,
+        # avoiding deadlocks when all threads are busy
+        async with sem:
+            yield await run_in_threadpool(cm.__enter__)
```

Comment:
Shouldn't thread be automatically closed, after completing its task?

Reply:
Seems reasonable to me!